lsp/src/main/kotlin/org/gradle/declarative/lsp/DeclarativeTextDocumentService.kt

```diff
@@ -297,20 +300,23 @@ class DeclarativeTextDocumentService : TextDocumentService {
     }
 
     private fun parse(uri: URI, text: String): DocumentOverlayResult {
+        fun AnalysisSequenceResult.lastStepDocument() =
+            stepResults.values.last().stepResultOrPartialResult.resolvedDocument()
+        
         val fileName = uri.path.substringAfterLast('/')
-        val fileSchema = schemaAnalysisEvaluator.evaluate(fileName, text)
-        val settingsSchema = schemaAnalysisEvaluator.evaluate(
-            declarativeResources.settingsFile.name, 
-            declarativeResources.settingsFile.takeIf { it.canRead() }?.readText().orEmpty()
-        )
-
-        val document = AnalysisDocumentUtils.documentWithModelDefaults(settingsSchema, fileSchema)
-        when (document != null) {
-            true -> LOGGER.trace("Parsed declarative model for document: {}", uri)
-            false -> LOGGER.error("Failed to parse declarative model for document: {}", uri)
-        }
-
-        return document!!
+        val document = schemaAnalysisEvaluator.evaluate(fileName, text).lastStepDocument()
+        
+        // Workaround: for now, the mutation utilities cannot handle mutations that touch the underlay document content.
+        // To avoid that, use an empty document as an underlay instead of the real document produced from the
+        // settings file.
+        // TODO: carry both the real overlay and the document produced from just the current file, run the mutations
+        //       against the latter for now.
+        // TODO: once the mutation utilities start handling mutations across the overlay, pass them the right overlay.
```

Comment:
🤔 Do you think issues would be useful for these, or are you going to come back here soon?

Reply:
I have some more changes in mind that are going to affect this.